src/server/game/Movement/MovementGenerators/ConfusedMovementGenerator.cpp

```diff
@@ -17,91 +17,94 @@
  */
 
 #include "Creature.h"
-#include "ConfusedMovementGenerator.h"
+#include "Player.h"
 #include "PathGenerator.h"
 #include "MoveSplineInit.h"
 #include "MoveSpline.h"
-#include "Player.h"
+#include "ConfusedMovementGenerator.h"
 
 template<class T>
-void ConfusedMovementGenerator<T>::DoInitialize(T* unit)
+ConfusedMovementGenerator<T>::~ConfusedMovementGenerator()
 {
-    unit->AddUnitState(UNIT_STATE_CONFUSED);
-    unit->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_CONFUSED);
-    unit->GetPosition(i_x, i_y, i_z);
+    delete _path;
+}
 
-    if (!unit->IsAlive() || unit->IsStopped())
+template<class T>
+void ConfusedMovementGenerator<T>::DoInitialize(T* owner)
+{
+    if (!owner || !owner->IsAlive())
         return;
 
-    unit->StopMoving();
-    unit->AddUnitState(UNIT_STATE_CONFUSED_MOVE);
+    owner->AddUnitState(UNIT_STATE_CONFUSED);
+    owner->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_CONFUSED);
+    owner->StopMoving();
+
+    _timer.Reset(0);
+    owner->GetPosition(_reference.x, _reference.y, _reference.z);
 }
 
 template<class T>
-void ConfusedMovementGenerator<T>::DoReset(T* unit)
+void ConfusedMovementGenerator<T>::DoReset(T* owner)
 {
-    i_nextMoveTime.Reset(0);
-
-    if (!unit->IsAlive() || unit->IsStopped())
-        return;
-
-    unit->StopMoving();
-    unit->AddUnitState(UNIT_STATE_CONFUSED | UNIT_STATE_CONFUSED_MOVE);
+    DoInitialize(owner);
 }
 
 template<class T>
-bool ConfusedMovementGenerator<T>::DoUpdate(T* unit, uint32 diff)
+bool ConfusedMovementGenerator<T>::DoUpdate(T* owner, uint32 diff)
 {
-    if (unit->HasUnitState(UNIT_STATE_ROOT | UNIT_STATE_STUNNED | UNIT_STATE_DISTRACTED))
-        return true;
+    if (!owner || !owner->IsAlive())
+        return false;
 
-    if (i_nextMoveTime.Passed())
+    if (owner->HasUnitState(UNIT_STATE_NOT_MOVE) || owner->IsMovementPreventedByCasting())
```

Comment:
Yes. That would be a good start in the right direction I think. At term though, it would be even better than the MoveGens do not call any methods of the unit or its AI. The AI should know the movement code and call it when it needs to move the unit its controlling. The movement code shouldn't be aware of the AI, nor of the unit as far as possible.

This decoupling will help us by making things simpler and easier to manage I think ([divide and conquer strategy](https://effectivesoftwaredesign.com/2011/06/06/divide-and-conquer-coping-with-complexity/)).